### PR TITLE
nixos/dhcpcd: harden and run as unprivileged user

### PR DESCRIPTION
**Pull request by rnhmjoj**
## Description of changes

These changes replace the dhcpcd privsep mode with a combination of POSIX capabilities and systemd security features that allow to fully run dhcpcd as an unprivileged user. See the commit messages for why I think this is an improvement.

There are a couple of backward incompatibilities, but most users shouldn't notice any difference.

## Things done

- [x] Tested via `dhcpcd.tests`
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bjornfor** [comment]
> nixos/dhcpcd: harden and run as unprivileged user 

Did you consider/try with DynamicUser?

**Comment by rnhmjoj** [comment]
> Did you consider/try with DynamicUser?

No, but I think it should work alright: there isn't too much state being handled (just a couple of files per interface + one for each wireless network) and no other user in the dhcpcd group.

**Comment by rnhmjoj** [comment]
@bjornfor I switched to DynamicUser.

File: nixos/modules/services/networking/dhcpcd.nix [hunk 1]
```diff
@@ -221,8 +221,19 @@ in
             Group = "dhcpcd";
             StateDirectory = "dhcpcd";
             RuntimeDirectory = "dhcpcd";
+
+            ExecStartPre = "+${pkgs.writeShellScript "migrate-dhcpcd" ''
+              # migrate from old database directory
+              if test -d /var/db/dhcpcd; then
+                mv /var/db/dhcpcd/* -t /var/lib/dhcpcd
+                rmdir /var/db/dhcpcd
+                chown -R dhcpcd:dhcpcd /var/lib/dhcpcd
```
Comment (bjornfor) [inline nixos/modules/services/networking/dhcpcd.nix:230]: Isn't this running as unprivileged user? Are you sure chmod works?

Comment (rnhmjoj) [reply nixos/modules/services/networking/dhcpcd.nix:230]: Yes, it works: the trick is the `+` at the beginning. (See "Table 2. Special executable prefixes" in `systemd.service(5)`)

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 2]
```diff
@@ -69,7 +69,10 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
   - If [`system.stateVersion`](#opt-system.stateVersion) is >=23.11, `pkgs.nextcloud27` will be installed by default.
   - Please note that an upgrade from v26 (or older) to v28 directly is not possible. Please upgrade to `nextcloud27` (or earlier) first. Nextcloud prohibits skipping major versions while upgrading. You can upgrade by declaring [`services.nextcloud.package = pkgs.nextcloud27;`](options.html#opt-services.nextcloud.package).
 
-- `services.resolved.fallbackDns` can now be used to disable the upstream fallback servers entirely by setting it to an empty list. To get the previous behaviour of the upstream defaults set it to null, the new default, instead.
+- The dhcpcd service (`networking.useDHCP`) has been hardened and now runs exclusively as the "dhcpcd" user.
```
Comment (bjornfor) [inline nixos/doc/manual/release-notes/rl-2405.section.md:72]: Is removing the `services.resolved.fallbackDns` entry the result of a bad merge/rebase?

Comment (rnhmjoj) [reply nixos/doc/manual/release-notes/rl-2405.section.md:72]: Oops, bad merge.

Comment (rnhmjoj) [reply nixos/doc/manual/release-notes/rl-2405.section.md:74]: Fixed.

File: nixos/modules/services/networking/dhcpcd.nix [hunk 3]
```diff
@@ -221,8 +221,19 @@ in
...
+              fi
+            ''}";
+
             ExecStart = "@${dhcpcd}/sbin/dhcpcd dhcpcd --quiet ${optionalString cfg.persistent "--persistent"} --config ${dhcpcdConf}";
             ExecReload = "${dhcpcd}/sbin/dhcpcd --rebind";
+
```
[the remark this answers is not in the document]
Comment (rnhmjoj) [reply nixos/modules/services/networking/dhcpcd.nix]: Fixed.

**Comment by rnhmjoj** [comment]
@grahamcOfBorg test dhcpcd.tests

**Review by Janik-Haag — COMMENTED** [review]
Changes look good to me, but I would prefer if someone more knowledgeable would look over this as well (I was only pinged because I recently started maintaining the networking tests).

While at it it would be cool if you could also add a dns test for `nixos/tests/networking/networkmanager.nix` that checks if this is working (if there isn't already one)

**Comment by rnhmjoj** [comment]
> While at it it would be cool if you could also add a dns test for nixos/tests/networking/networkmanager.nix that checks if this is working (if there isn't already one)

I'm not familiar with NetworkManager, but I think it doesn't use dhcpcd at all. At least, in the test you mentioned it's being explicitly disabled.

**Comment by uninsane** [comment]
NetworkManager defaults to its own internal dhcp client, but does expose a config option to instead use dhcpcd. in nixos, that's:
```nix
networking.networkmanager.dhcp = "dhcpcd";
```
https://github.com/NixOS/nixpkgs/blob/e1c84c556c9f8e1983508f26198fc358a3f4ed50/nixos/modules/services/networking/networkmanager.nix#L225-L231

**Pull request closed without merging**